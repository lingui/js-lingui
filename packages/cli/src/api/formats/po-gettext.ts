import { format as formatDate } from "date-fns"
import fs from "fs"
import ICUParser from "messageformat-parser"
import pluralsCldr from "plurals-cldr"
import PO from "pofile"
import * as R from "ramda"
import gettextPlurals from "node-gettext/lib/plurals"

import { CatalogType, MessageType } from "../catalog"
import { joinOrigin, splitOrigin, writeFileIfChanged } from "../utils"
import { CatalogFormatter } from "."

// Workaround because pofile doesn't support es6 modules, see https://github.com/rubenv/pofile/pull/38#issuecomment-623119284
type POItemType = InstanceType<typeof PO.Item>

function getCreateHeaders(language = "no") {
  return {
    "POT-Creation-Date": formatDate(new Date(), "yyyy-MM-dd HH:mmxxxx"),
    "MIME-Version": "1.0",
    "Content-Type": "text/plain; charset=utf-8",
    "Content-Transfer-Encoding": "8bit",
    "X-Generator": "@lingui/cli",
    Language: language,
  }
}

// Attempts to turn a single tokenized ICU plural case back into a string.
function stringifyICUCase(icuCase) {
  return icuCase.tokens
    .map((token) => {
      if (typeof token === "string") {
        return token
      } else if (token.type === "octothorpe") {
        return "#"
      } else if (token.type === "argument") {
        return "{" + token.arg + "}"
      } else {
        console.warn(
          `Unexpected token "${token}" while stringifying plural case "${icuCase}". Token will be ignored.`
        )
        return ""
      }
    })
    .join("")
}

const ICU_PLURAL_REGEX = /^{.*, plural, .*}$/
const ICU_SELECT_REGEX = /^{.*, select(Ordinal)?, .*}$/
const LINE_ENDINGS = /\r?\n/g

// Prefix that is used to identitify context information used by this module in PO's "extracted comments".
const CTX_PREFIX = "js-lingui:"

const serialize = (items: CatalogType, options) =>
  R.compose(
    R.values,
    R.mapObjIndexed((message: MessageType, key) => {
      const item = new PO.Item()
      item.msgid = key
      item.comments = message.comments || []

      // The extractedComments array may be modified in this method, so create a new array with the message's elements.
      // Destructuring `undefined` is forbidden, so fallback to `[]` if the message has no extracted comments.
      item.extractedComments = [...(message.extractedComments ?? [])]
      
      if (options.origins !== false) {
        if (message.origin && options.lineNumbers === false) {
          item.references = message.origin.map(msg => msg.slice(0, -1)).map(joinOrigin)
        } else {
          item.references = message.origin ? message.origin.map(joinOrigin) : []
        }
      }

      // @ts-ignore: Figure out how to set this flag
      item.obsolete = message.obsolete
      item.flags = message.flags
        ? R.fromPairs(message.flags.map((flag) => [flag, true]))
        : {}

      // Depending on whether custom ids are used by the developer, the (potential plural) "original", untranslated ICU
      // message can be found in `message.message` or in the item's `key` itself.
      const icuMessage = message.message || key

      const _simplifiedMessage = icuMessage.replace(LINE_ENDINGS, " ")

      // Quick check to see if original message is a plural localization.
      if (ICU_PLURAL_REGEX.test(_simplifiedMessage)) {
        try {
          const [messageAst] = ICUParser.parse(icuMessage)

          // Check if any of the plural cases contain plurals themselves.
          if (
            messageAst.cases.some((icuCase) =>
              icuCase.tokens.some((token) => token.type === "plural")
            )
          ) {
            console.warn(
              `Nested plurals cannot be expressed with gettext plurals. ` +
                `Message with key "%s" will not be saved correctly.`,
              key
            )
          }

          // Store placeholder that is pluralized upon to allow restoring ICU format later.
          const ctx = new URLSearchParams({
            pluralize_on: messageAst.arg,
          })

          if (message.message == null) {
            // For messages without developer-set ID, use first case as `msgid` and the last case as `msgid_plural`.
            // This does not necessarily make sense for development languages with more than two numbers, but gettext
            // only supports exactly two plural forms.
            item.msgid = stringifyICUCase(messageAst.cases[0])
            item.msgid_plural = stringifyICUCase(
              messageAst.cases[messageAst.cases.length - 1]
            )

            // Since the original msgid is overwritten, store ICU message to allow restoring that ID later.
            ctx.set("icu", key)
          } else {
            // For messages with developer-set ID, append `_plural` to the key to generate `msgid_plural`.
            item.msgid_plural = key + "_plural"
          }

          ctx.sort()
          item.extractedComments.push(CTX_PREFIX + ctx.toString())

          // If there is a translated value, parse that instead of the original message to prevent overriding localized
          // content with the original message. If there is no translated value, don't touch msgstr, since marking item as
          // plural (above) already causes `pofile` to automatically generate `msgstr[0]` and `msgstr[1]`.
          if (message.translation?.length > 0) {
            try {
              const [ast] = ICUParser.parse(message.translation)
              if (ast.cases == null) {
                console.warn(
                  `Found translation without plural cases for key "${key}". ` +
                    `This likely means that a translated .po file misses multiple msgstr[] entries for the key. ` +
                    `Translation found: "${message.translation}"`
                )
                item.msgstr = [message.translation]
              } else {
                item.msgstr = ast.cases.map(stringifyICUCase)
              }
            } catch (e) {
              console.error(`Error parsing translation ICU for key "${key}"`, e)
            }
          }
        } catch (e) {
          console.error(`Error parsing message ICU for key "${key}":`, e)
        }
      } else {
        if (
          !options.disableSelectWarning &&
          ICU_SELECT_REGEX.test(_simplifiedMessage)
        ) {
          console.warn(
            `ICU 'select' and 'selectOrdinal' formats cannot be expressed natively in gettext format. ` +
              `Item with key "%s" will be included in the catalog as raw ICU message. ` +
              `To disable this warning, include '{ disableSelectWarning: true }' in the config's 'formatOptions'`,
            key
          )
        }
        item.msgstr = [message.translation]
      }

      return item
    })
  )(items)

const getMessageKey = R.prop<"msgid", string>("msgid")
const getTranslations = R.prop<"msgstr", string[]>("msgstr")
const getExtractedComments = R.prop("extractedComments")
const getTranslatorComments = R.prop("comments")
const getOrigins = R.prop("references")
const getFlags = R.compose(
  R.map(R.trim),
  R.keys,
  R.dissoc("obsolete"), // backward-compatibility, remove in 3.x
  R.prop("flags")
)
const isObsolete = R.either(R.path(["flags", "obsolete"]), R.prop("obsolete"))

const getTranslationCount = R.compose(R.length, getTranslations)

const deserialize: (Object) => Object = R.map(
  R.applySpec({
    translation: R.compose(R.head, R.defaultTo([]), getTranslations),
    extractedComments: R.compose(R.defaultTo([]), getExtractedComments),
    comments: R.compose(R.defaultTo([]), getTranslatorComments),
    obsolete: isObsolete,
    origin: R.compose(R.map(splitOrigin), R.defaultTo([]), getOrigins),
    flags: getFlags,
  })
)

/**
 * Returns ICU case labels in the order that gettext lists localized messages, e.g. 0,1,2 => `["one", "two", "other"]`.
 *
 * Obtaining the ICU case labels for gettext-supported inputs (gettext doesn't support fractions, even though some
 * languages have a separate case for fractional numbers) works by applying the CLDR selector to the example values
 * listed in the node-gettext plurals module.
 *
 * This approach is heavily influenced by
 * https://github.com/LLK/po2icu/blob/9eb97f81f72b2fee02b77f1424702e019647e9b9/lib/po2icu.js#L148.
 */
const getPluralCases = (lang: string): string[] | undefined => {
  const gettextPluralsInfo = gettextPlurals[lang]

  return gettextPluralsInfo?.examples.map((pluralCase) =>
    pluralsCldr(lang, pluralCase.sample)
  )
}

const convertPluralsToICU = (items: POItemType[], lang?: string) => {
  // .po plurals are numbered 0-N and need to be mapped to ICU plural classes ("one", "few", "many"...). Different
  // languages can have different plural classes (some start with "zero", some with "one"), so read that data from CLDR.
  // `pluralForms` may be `null` if lang is not found. As long as no plural is used, don't report an error.
  let pluralForms = getPluralCases(lang)

  items.forEach((item) => {
    const translationCount = getTranslationCount(item)
    const messageKey = getMessageKey(item)

    // Messages without multiple translations (= plural cases) need no further processing.
    if (translationCount <= 1) {
      return
    }

    // msgid_plural must be set, but its actual value is not important.
    if (!item.msgid_plural) {
      console.warn(
        `Multiple translations for item with key "%s" but missing 'msgid_plural' in catalog "${lang}". This is not supported and the plural cases will be ignored.`,
        messageKey
      )
      return
    }

    const contextComment = item.extractedComments
      .find((comment) => comment.startsWith(CTX_PREFIX))
      ?.substr(CTX_PREFIX.length)
    const ctx = new URLSearchParams(contextComment)

    if (contextComment != null) {
      item.extractedComments = item.extractedComments.filter(
        (comment) => !comment.startsWith(CTX_PREFIX)
      )
    }

    // If an original ICU was stored, use that as `msgid` to match the catalog that was originally exported.
    const storedICU = ctx.get("icu")
    if (storedICU != null) {
      item.msgid = storedICU
    }

    // If all translations are empty, ignore item.
    if (item.msgstr.every((str) => str.length === 0)) {
      return
    }

    if (pluralForms == null) {
      console.warn(
        `Multiple translations for item with key "%s" in language "${lang}", but no plural cases were found. ` +
          `This prohibits the translation of .po plurals into ICU plurals. Pluralization will not work for this key.`,
        messageKey
      )
      return
    }

    const pluralCount = pluralForms.length
    if (translationCount > pluralCount) {
      console.warn(
        `More translations provided (${translationCount}) for item with key "%s" in language "${lang}" than there are plural cases available (${pluralCount}). ` +
          `This will result in not all translations getting picked up.`,
        messageKey
      )
    }

    // Map each msgstr to a "<pluralform> {<translated_string>}" entry, joined by one space.
    const pluralClauses = item.msgstr
      .map((str, index) => pluralForms[index] + " {" + str + "}")
      .join(" ")

    // Find out placeholder name from item's message context, defaulting to "count".
    let pluralizeOn = ctx.get("pluralize_on")
    if (!pluralizeOn) {
      console.warn(
        `Unable to determine plural placeholder name for item with key "%s" in language "${lang}" (should be stored in a comment starting with "#. ${CTX_PREFIX}"), assuming "count".`,
        messageKey
      )
      pluralizeOn = "count"
    }

    item.msgstr = ["{" + pluralizeOn + ", plural, " + pluralClauses + "}"]
  })
}

const indexItems = R.indexBy(getMessageKey)

type PoFormatter = {
  parse: (raw: string) => Object
  serialize: (
    catalog: CatalogType,
    options: Record<string, any>
  ) => POItemType[]
}

const poGettext: CatalogFormatter & PoFormatter = {
  catalogExtension: ".po",

  write(filename, catalog: CatalogType, options) {
    let po
    if (fs.existsSync(filename)) {
      const raw = fs.readFileSync(filename).toString()
      po = PO.parse(raw)
    } else {
      po = new PO()
      po.headers = getCreateHeaders(options.locale)
      if (options.locale === undefined) {
        delete po.headers.Language
      }
      po.headerOrder = Object.keys(po.headers)
    }
    po.items = this.serialize(catalog, options)
    writeFileIfChanged(filename, po.toString())
  },

  // Mainly exported for easier testing
  serialize,

  read(filename) {
    const raw = fs.readFileSync(filename).toString()
    return this.parse(raw)
  },

  parse(raw: string) {
    const po = PO.parse(raw)
    convertPluralsToICU(po.items, po.headers.Language)
    return deserialize(indexItems(po.items))
  },
}

export default poGettext
