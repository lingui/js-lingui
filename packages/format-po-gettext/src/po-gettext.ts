import { parse as parseIcu, Select, SelectCase } from "@messageformat/parser"
import pluralsCldr from "plurals-cldr"
import PO from "pofile"
// @ts-ignore
import gettextPlurals from "node-gettext/lib/plurals"

import type { CatalogFormatter, CatalogType, MessageType } from "@lingui/conf"
import { generateMessageId } from "@lingui/message-utils/generateMessageId"
import { formatter as poFormatter } from "@lingui/format-po"
import type { PoFormatterOptions } from "@lingui/format-po"

type POItem = InstanceType<typeof PO.Item>

export type PoGettextFormatterOptions = PoFormatterOptions & {
  disableSelectWarning?: boolean
}

// Attempts to turn a single tokenized ICU plural case back into a string.
function stringifyICUCase(icuCase: SelectCase): string {
  return icuCase.tokens
    .map((token) => {
      if (token.type === "content") {
        return token.value
      } else if (token.type === "octothorpe") {
        return "#"
      } else if (token.type === "argument") {
        return "{" + token.arg + "}"
      } else {
        console.warn(
          `Unexpected token "${token}" while stringifying plural case "${icuCase}". Token will be ignored.`
        )
        return ""
      }
    })
    .join("")
}

const ICU_PLURAL_REGEX = /^{.*, plural, .*}$/
const ICU_SELECT_REGEX = /^{.*, select(Ordinal)?, .*}$/
const LINE_ENDINGS = /\r?\n/g

// Prefix that is used to identitify context information used by this module in PO's "extracted comments".
const CTX_PREFIX = "js-lingui:"

function serializePlurals(
  item: POItem,
  message: MessageType,
  id: string,
  isGeneratedId: boolean,
  options: PoGettextFormatterOptions
): POItem {
  // Depending on whether custom ids are used by the developer, the (potential plural) "original", untranslated ICU
  // message can be found in `message.message` or in the item's `key` itself.
  const icuMessage = message.message

  if (!icuMessage) {
    return item
  }

  const _simplifiedMessage = icuMessage.replace(LINE_ENDINGS, " ")

  // Quick check to see if original message is a plural localization.
  if (ICU_PLURAL_REGEX.test(_simplifiedMessage)) {
    try {
      const messageAst = parseIcu(icuMessage)[0] as Select

      // Check if any of the plural cases contain plurals themselves.
      if (
        messageAst.cases.some((icuCase) =>
          icuCase.tokens.some((token) => token.type === "plural")
        )
      ) {
        console.warn(
          `Nested plurals cannot be expressed with gettext plurals. ` +
            `Message with key "%s" will not be saved correctly.`,
          id
        )
      }

      // Store placeholder that is pluralized upon to allow restoring ICU format later.
      const ctx = new URLSearchParams({
        pluralize_on: messageAst.arg,
      })

      if (isGeneratedId) {
        // For messages without developer-set ID, use first case as `msgid` and the last case as `msgid_plural`.
        // This does not necessarily make sense for development languages with more than two numbers, but gettext
        // only supports exactly two plural forms.
        item.msgid = stringifyICUCase(messageAst.cases[0])
        item.msgid_plural = stringifyICUCase(
          messageAst.cases[messageAst.cases.length - 1]
        )

        // Since the original msgid is overwritten, store ICU message to allow restoring that ID later.
        ctx.set("icu", icuMessage)
      } else {
        // For messages with developer-set ID, append `_plural` to the key to generate `msgid_plural`.
        item.msgid_plural = id + "_plural"
      }

      ctx.sort()
      item.extractedComments.push(CTX_PREFIX + ctx.toString())

      // If there is a translated value, parse that instead of the original message to prevent overriding localized
      // content with the original message. If there is no translated value, don't touch msgstr, since marking item as
      // plural (above) already causes `pofile` to automatically generate `msgstr[0]` and `msgstr[1]`.
      if (message.translation?.length > 0) {
        const ast = parseIcu(message.translation)[0] as Select
        if (ast.cases == null) {
          console.warn(
            `Found translation without plural cases for key "${id}". ` +
              `This likely means that a translated .po file misses multiple msgstr[] entries for the key. ` +
              `Translation found: "${message.translation}"`
          )
          item.msgstr = [message.translation]
        } else {
          item.msgstr = ast.cases.map(stringifyICUCase)
        }
      }
    } catch (e) {
      console.error(`Error parsing message ICU for key "${id}":`, e)
    }
  } else {
    if (
      !options.disableSelectWarning &&
      ICU_SELECT_REGEX.test(_simplifiedMessage)
    ) {
      console.warn(
        `ICU 'select' and 'selectOrdinal' formats cannot be expressed natively in gettext format. ` +
          `Item with key "%s" will be included in the catalog as raw ICU message. ` +
          `To disable this warning, include '{ disableSelectWarning: true }' in the config's 'formatOptions'`,
        id
      )
    }
    item.msgstr = [message.translation]
  }

  return item
}

/**
 * Returns ICU case labels in the order that gettext lists localized messages, e.g. 0,1,2 => `["one", "two", "other"]`.
 *
 * Obtaining the ICU case labels for gettext-supported inputs (gettext doesn't support fractions, even though some
 * languages have a separate case for fractional numbers) works by applying the CLDR selector to the example values
 * listed in the node-gettext plurals module.
 *
 * This approach is heavily influenced by
 * https://github.com/LLK/po2icu/blob/9eb97f81f72b2fee02b77f1424702e019647e9b9/lib/po2icu.js#L148.
 */
const getPluralCases = (lang: string): string[] | undefined => {
  // If users uses locale with underscore or slash, es-ES, es_ES, gettextplural is "es" not es-ES.
  const [correctLang] = lang.split(/[-_]/g)
  const gettextPluralsInfo = gettextPlurals[correctLang]

  return gettextPluralsInfo?.examples.map((pluralCase: any) =>
    pluralsCldr(correctLang, pluralCase.sample)
  )
}

const convertPluralsToICU = (
  item: POItem,
  pluralForms: string[],
  lang: string
) => {
  const translationCount = item.msgstr.length
  const messageKey = item.msgid

  // Messages without multiple translations (= plural cases) need no further processing.
  if (translationCount <= 1 && !item.msgid_plural) {
    return
  }

  // msgid_plural must be set, but its actual value is not important.
  if (!item.msgid_plural) {
    console.warn(
      `Multiple translations for item with key "%s" but missing 'msgid_plural' in catalog "${lang}". This is not supported and the plural cases will be ignored.`,
      messageKey
    )
    return
  }

  const contextComment = item.extractedComments
    .find((comment) => comment.startsWith(CTX_PREFIX))
    ?.substr(CTX_PREFIX.length)
  const ctx = new URLSearchParams(contextComment)

  if (contextComment != null) {
    item.extractedComments = item.extractedComments.filter(
      (comment) => !comment.startsWith(CTX_PREFIX)
    )
  }

  // If an original ICU was stored, use that as `msgid` to match the catalog that was originally exported.
  const storedICU = ctx.get("icu")
  if (storedICU != null) {
    item.msgid = storedICU
  }

  // If all translations are empty, ignore item.
  if (item.msgstr.every((str) => str.length === 0)) {
    return
  }

  if (pluralForms == null) {
    console.warn(
      `Multiple translations for item with key "%s" in language "${lang}", but no plural cases were found. ` +
        `This prohibits the translation of .po plurals into ICU plurals. Pluralization will not work for this key.`,
      messageKey
    )
    return
  }

  const pluralCount = pluralForms.length
  if (translationCount > pluralCount) {
    console.warn(
      `More translations provided (${translationCount}) for item with key "%s" in language "${lang}" than there are plural cases available (${pluralCount}). ` +
        `This will result in not all translations getting picked up.`,
      messageKey
    )
  }

  // Map each msgstr to a "<pluralform> {<translated_string>}" entry, joined by one space.
  const pluralClauses = item.msgstr
    .map((str, index) => pluralForms[index] + " {" + str + "}")
    .join(" ")

  // Find out placeholder name from item's message context, defaulting to "count".
  let pluralizeOn = ctx.get("pluralize_on")
  if (!pluralizeOn) {
    console.warn(
      `Unable to determine plural placeholder name for item with key "%s" in language "${lang}" (should be stored in a comment starting with "#. ${CTX_PREFIX}"), assuming "count".`,
      messageKey
    )
    pluralizeOn = "count"
  }

  item.msgstr = ["{" + pluralizeOn + ", plural, " + pluralClauses + "}"]
}

export function formatter(
  options: PoGettextFormatterOptions = {}
): CatalogFormatter {
  options = {
    origins: true,
    lineNumbers: true,
    ...options,
  }

  const formatter = poFormatter(options)

  return {
    catalogExtension: ".po",
    templateExtension: ".pot",

    parse(content, ctx): CatalogType {
      const po = PO.parse(content)

      // .po plurals are numbered 0-N and need to be mapped to ICU plural classes ("one", "few", "many"...). Different
      // languages can have different plural classes (some start with "zero", some with "one"), so read that data from CLDR.
      // `pluralForms` may be `null` if lang is not found. As long as no plural is used, don't report an error.
      let pluralForms = getPluralCases(po.headers.Language)

      po.items.forEach((item) => {
        convertPluralsToICU(item, pluralForms, po.headers.Language)
      })

      return formatter.parse(po.toString(), ctx) as CatalogType
    },

    serialize(catalog, ctx): string {
      const po = PO.parse(formatter.serialize(catalog, ctx) as string)

      po.items = po.items.map((item) => {
        const isGeneratedId = !item.extractedComments.includes(
          "js-lingui-explicit-id"
        )
        const id = isGeneratedId
          ? generateMessageId(item.msgid, item.msgctxt)
          : item.msgid
        const message = catalog[id]
        return serializePlurals(item, message, id, isGeneratedId, options)
      })

      return po.toString()
    },
  }
}
