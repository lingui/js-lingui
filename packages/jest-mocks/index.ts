export function getConsoleMockCalls({ mock }: jest.MockInstance<any, any>) {
  if (!mock.calls.length) return
  return mock.calls.map((call) => call[0]).join("\n")
}

export function mockConsole(
  testCase: (console: jest.Mocked<Console>) => any,
  mock = {}
) {
  function restoreConsole() {
    global.console = originalConsole
  }

  const originalConsole = global.console

  const defaults = {
    log: jest.fn(),
    warn: jest.fn(),
    error: jest.fn(),
  }

  global.console = {
    ...defaults,
    ...mock,
  } as any

  let result
  try {
    result = testCase(global.console as jest.Mocked<Console>)
  } catch (e) {
    restoreConsole()
    throw e
  }

  if (result && typeof result.then === "function") {
    return result.then(restoreConsole).catch((e) => {
      restoreConsole()
      throw e
    })
  } else {
    restoreConsole()
    return result
  }
}

export function mockEnv(env, testCase) {
  const oldEnv = process.env.NODE_ENV
  process.env.NODE_ENV = env

  try {
    testCase()
  } catch (e) {
    process.env.NODE_ENV = oldEnv
    throw e
  }

  process.env.NODE_ENV = oldEnv
}
