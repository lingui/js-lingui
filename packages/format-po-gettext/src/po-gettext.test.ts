import { mockConsole } from "@lingui/jest-mocks"
import fs from "fs"
import path from "path"

import { CatalogFormatter, CatalogType } from "@lingui/conf"
import { formatter as createFormat } from "./po-gettext"
import MockDate from "mockdate"

const defaultParseCtx: Parameters<CatalogFormatter["parse"]>[1] = {
  locale: "en",
  sourceLocale: "en",
  filename: "file.po",
}

const defaultSerializeCtx: Parameters<CatalogFormatter["serialize"]>[1] = {
  locale: "en",
  existing: null,
  filename: "file.po",
  sourceLocale: "en",
}

describe("po-gettext format", () => {
  const format = createFormat()

  beforeAll(() => {
    MockDate.set(new Date("2018-08-27T10:00Z"))
  })

  afterAll(() => {
    MockDate.reset()
  })

  it("should convert ICU plural messages to gettext plurals", () => {
    const catalog: CatalogType = {
      static: {
        translation: "Static message",
      },
      message_with_id_and_octothorpe: {
        message: "{count, plural, one {Singular} other {Number is #}}",
        translation: "{count, plural, one {Singular} other {Number is #}}",
        comments: ["js-lingui-explicit-id"],
      },
      message_with_id: {
        message:
          "{someCount, plural, one {Singular case with id\
          and linebreak} other {Case number {someCount} with id}}",
        translation:
          "{someCount, plural, one {Singular case with id} other {Case number {someCount} with id}}",
        comments: [
          "This is a comment by the developers about how the content must be localized.",
          "js-lingui-explicit-id",
        ],
      },
      WGI12K: {
        message:
          "{anotherCount, plural, one {Singular case} other {Case number {anotherCount}}}",
        translation:
          "{anotherCount, plural, one {Singular case} other {Case number {anotherCount}}}",
      },
      // Entry with developer-defined ID that generates empty msgstr[] lines
      message_with_id_but_without_translation: {
        message:
          "{count, plural, one {Singular with id but no translation} other {Plural {count} with empty id but no translation}}",
        translation: "",
        comments: ["js-lingui-explicit-id"],
      },
      // Entry with automatic ID that generates empty msgstr[] lines
      xZCXAV: {
        message:
          "{count, plural, one {Singular automatic id no translation} other {Plural {count} automatic id no translation}}",
        translation: "",
      },
    }

    const pofile = format.serialize(catalog, defaultSerializeCtx)

    expect(pofile).toMatchSnapshot()
  })

  it("should convert gettext plurals to ICU plural messages", () => {
    const pofile = fs
      .readFileSync(path.join(__dirname, "fixtures/messages_plural.po"))
      .toString()

    const catalog = format.parse(pofile, defaultParseCtx)

    expect(catalog).toMatchSnapshot()
  })

  it("should warn when using nested plurals that cannot be represented with gettext plurals", () => {
    const catalog = {
      nested_plural_message: {
        message: `{count, plural,
          one {{numArticles, plural,
            one {1 book and 1 article}
            other {1 book and {numArticles} articles}
          }}
          other {{numArticles, plural,
            one {{numBooks} books and 1 article}
            other {{numBooks} books and {numArticles} articles}
          }}
        }`,
        translation: `{count, plural,
          one {{numArticles, plural,
            one {1 book and 1 article}
            other {1 book and {numArticles} articles}
          }}
          other {{numArticles, plural,
            one {{numBooks} books and 1 article}
            other {{numBooks} books and {numArticles} articles}
          }}
        }`,
      },
    }

    mockConsole((console) => {
      format.serialize(catalog, defaultSerializeCtx)

      expect(console.warn).toHaveBeenCalledWith(
        expect.stringContaining("Nested plurals"),
        "nested_plural_message"
      )
    })
  })

  it("should use correct ICU plural cases for languages having an additional plural case for fractions", () => {
    // This tests the edge case described in https://github.com/lingui/js-lingui/pull/677#issuecomment-737152022
    const po = `
msgid ""
msgstr ""
"Language: cs\n"

#. js-lingui:icu=%7B#%2C+plural%2C+one+%7Bday%7D+other+%7Bdays%7D%7D&pluralize_on=#
msgid "# day"
msgid_plural "# days"
msgstr[0] "# den"
msgstr[1] "# dny"
msgstr[2] "# dní"
  `

    const parsed = format.parse(po, defaultParseCtx)

    // Note that the last case must be `other` (the 4th CLDR case name) instead of `many` (the 3rd CLDR case name).
    expect(parsed).toMatchInlineSnapshot(`
      {
        Y8Xw2Y: {
          comments: [],
          context: null,
          extra: {
            flags: [],
            translatorComments: [],
          },
          message: {#, plural, one {day} other {days}},
          obsolete: false,
          origin: [],
          translation: {#, plural, one {# den} few {# dny} other {# dní}},
        },
      }
    `)
  })

  describe("when using 'select' format", () => {
    const catalog = {
      select_message: {
        message: `{gender, select, male {he} female {she} other {they}`,
        translation: "",
      },
    }

    it("should warn", () => {
      mockConsole((console) => {
        format.serialize(catalog, defaultSerializeCtx)

        expect(console.warn).toHaveBeenCalledWith(
          expect.stringContaining("select"),
          "select_message"
        )
      })
    })

    it("should not warn when disabling the warning in config", () => {
      const format = createFormat({ disableSelectWarning: true })

      mockConsole((console) => {
        format.serialize(catalog, defaultSerializeCtx)

        expect(console.warn).not.toHaveBeenCalled()
      })
    })
  })

  describe("when using 'selectOrdinal' format", () => {
    const catalog = {
      select_ordinal_message: {
        message: `{count, selectOrdinal, one {#st} two {#nd} few {#rd} other {#th}}`,
        translation: "",
      },
    }

    it("should warn", () => {
      mockConsole((console) => {
        format.serialize(catalog, defaultSerializeCtx)

        expect(console.warn).toHaveBeenCalledWith(
          expect.stringContaining("selectOrdinal"),
          "select_ordinal_message"
        )
      })
    })

    it("should not warn when disabling the warning in config", () => {
      const format = createFormat({ disableSelectWarning: true })

      mockConsole((console) => {
        format.serialize(catalog, defaultSerializeCtx)

        expect(console.warn).not.toHaveBeenCalled()
      })
    })
  })

  it("convertPluralsToIco handle correctly locales with 4-letter", () => {
    const pofile = fs
      .readFileSync(
        path.join(__dirname, "fixtures/messages_plural-4-letter.po")
      )
      .toString()

    const catalog = format.parse(pofile, defaultParseCtx)

    expect(catalog).toMatchSnapshot()
  })
})
