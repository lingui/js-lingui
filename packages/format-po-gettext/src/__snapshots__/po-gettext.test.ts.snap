// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`po-gettext format convertPluralsToIco handle correctly locales with 4-letter 1`] = `
{
  WGI12K: {
    comments: [],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    message: {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}},
    obsolete: false,
    origin: [],
    translation: {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}},
  },
  jO/SBZ: {
    comments: [],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    message: {count, plural, one {Singular} other {Plural}},
    obsolete: false,
    origin: [],
    translation: ,
  },
  message_with_id: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: {someCount, plural, one {Singular case} other {Case number {someCount}}},
  },
  message_with_id_but_without_translation: {
    comments: [
      Comment made by the developers.,
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: ,
  },
}
`;

exports[`po-gettext format should convert ICU plural messages to gettext plurals 1`] = `
msgid ""
msgstr ""
"POT-Creation-Date: 2018-08-27 10:00+0000\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=utf-8\\n"
"Content-Transfer-Encoding: 8bit\\n"
"X-Generator: @lingui/cli\\n"
"Language: en\\n"
"Project-Id-Version: \\n"
"Report-Msgid-Bugs-To: \\n"
"PO-Revision-Date: \\n"
"Last-Translator: \\n"
"Language-Team: \\n"
"Plural-Forms: \\n"

#. js-lingui-explicit-id
msgid "static"
msgstr "Static message"

#. js-lingui-explicit-id
#. js-lingui:pluralize_on=count
msgid "message_with_id_and_octothorpe"
msgid_plural "message_with_id_and_octothorpe_plural"
msgstr[0] "Singular"
msgstr[1] "Number is #"

#. This is a comment by the developers about how the content must be localized.
#. js-lingui-explicit-id
#. js-lingui:pluralize_on=someCount
msgid "message_with_id"
msgid_plural "message_with_id_plural"
msgstr[0] "Singular case with id"
msgstr[1] "Case number {someCount} with id"

#. js-lingui:icu=%7BanotherCount%2C+plural%2C+one+%7BSingular+case%7D+other+%7BCase+number+%7BanotherCount%7D%7D%7D&pluralize_on=anotherCount
msgid "Singular case"
msgid_plural "Case number {anotherCount}"
msgstr[0] "Singular case"
msgstr[1] "Case number {anotherCount}"

#. js-lingui-explicit-id
#. js-lingui:pluralize_on=count
msgid "message_with_id_but_without_translation"
msgid_plural "message_with_id_but_without_translation_plural"
msgstr[0] ""
msgstr[1] ""

#. js-lingui:icu=%7Bcount%2C+plural%2C+one+%7BSingular+automatic+id+no+translation%7D+other+%7BPlural+%7Bcount%7D+automatic+id+no+translation%7D%7D&pluralize_on=count
msgid "Singular automatic id no translation"
msgid_plural "Plural {count} automatic id no translation"
msgstr[0] ""
msgstr[1] ""

`;

exports[`po-gettext format should convert gettext plurals to ICU plural messages 1`] = `
{
  WGI12K: {
    comments: [],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    message: {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}},
    obsolete: false,
    origin: [],
    translation: {anotherCount, plural, one {Singular case} other {Case number {anotherCount}}},
  },
  jO/SBZ: {
    comments: [],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    message: {count, plural, one {Singular} other {Plural}},
    obsolete: false,
    origin: [],
    translation: ,
  },
  message_with_id: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: {someCount, plural, one {Singular case} other {Case number {someCount}}},
  },
  message_with_id_but_without_translation: {
    comments: [
      Comment made by the developers.,
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: ,
  },
  static: {
    comments: [
      js-lingui-explicit-id,
    ],
    context: null,
    extra: {
      flags: [],
      translatorComments: [],
    },
    obsolete: false,
    origin: [],
    translation: Static message,
  },
}
`;
